ntpd/src/daemon/keyexchange.rs

```diff
@@ -647,4 +660,243 @@ mod tests {
+        let keyset = provider.get();
+
+        BoundKeyExchangeServer::run(stream, config, keyset, pool_certs).await
+    }
+
+    async fn client_tls_stream(
+        server_name: &str,
+        port: u16,
+    ) -> tokio_rustls::client::TlsStream<tokio::net::TcpStream> {
+        let stream = tokio::net::TcpStream::connect((server_name, port))
+            .await
+            .unwrap();
+
+        let ca = include_bytes!("../../test-keys/testca.pem");
+        let extra_certificates = &certificates_from_bufread(BufReader::new(Cursor::new(ca)));
+
+        let config = build_client_config(extra_certificates).unwrap();
+
+        let domain = rustls::ServerName::try_from(server_name)
+            .map_err(|_| std::io::Error::new(std::io::ErrorKind::InvalidInput, "invalid dnsname"))
+            .unwrap();
+
+        let connector = tokio_rustls::TlsConnector::from(Arc::new(config));
+        connector.connect(domain, stream).await.unwrap()
+    }
+
+    async fn send_records_to_server(records: Vec<NtsRecord>) -> Result<(), KeyExchangeError> {
+        let listener = TcpListener::bind(&("localhost", 0)).await?;
+        let port = listener.local_addr()?.port();
+
+        tokio::spawn(async move {
+            let mut stream = client_tls_stream("localhost", port).await;
+
+            for record in records {
+                let mut buffer = Vec::with_capacity(1024);
+                record.write(&mut buffer).unwrap();
+
+                stream.write_all(&buffer).await.unwrap();
+            }
+
+            let mut buf = [0; 1024];
+            loop {
+                match stream.read(&mut buf).await.unwrap() {
+                    0 => break,
+                    _ => continue,
+                }
+            }
+        });
+
+        run_server(listener).await
+    }
+
+    #[tokio::test]
+    async fn receive_cookies() {
+        let result = send_records_to_client(vec![
+            NtsRecord::NextProtocol {
+                protocol_ids: vec![0],
+            },
+            NtsRecord::AeadAlgorithm {
+                critical: false,
+                algorithm_ids: vec![15],
+            },
+            NtsRecord::NewCookie {
+                cookie_data: vec![1, 2, 3],
+            },
+            NtsRecord::EndOfMessage,
+        ])
+        .await;
+
+        assert!(result.is_ok());
+    }
+
+    #[tokio::test]
+    async fn records_after_end_are_ignored() {
+        let result = send_records_to_client(vec![
+            NtsRecord::NextProtocol {
+                protocol_ids: vec![0],
+            },
+            NtsRecord::AeadAlgorithm {
+                critical: false,
+                algorithm_ids: vec![15],
+            },
+            NtsRecord::NewCookie {
+                cookie_data: vec![1, 2, 3],
+            },
+            NtsRecord::EndOfMessage,
+            NtsRecord::NewCookie {
+                cookie_data: vec![1, 2, 3],
+            },
+        ])
+        .await;
+
+        assert!(result.is_ok());
+    }
```

Comment:
I believe this is is technically not spec-compliant 

> The sequence SHALL be terminated by a "End of Message" record

but we can interpret that in different ways (e.g. we do stop reading the sequence when we find the first end of message). Not a problem for valid clients.